Sources/MapboxMaps/Style/StyleSourceManager.swift

```diff
@@ -0,0 +1,180 @@
+
+        try setSourceProperties(for: id, properties: volatileProperties)
+    }
+
+    internal func updateGeoJSONSource(withId id: String, geoJSON: GeoJSONObject) throws {
+        guard let sourceInfo = allSourceIdentifiers.first(where: { $0.id == id }),
+              sourceInfo.type == .geoJson else {
+            throw StyleError(message: "Source with id '\(id)' is not found or not a GeoJSONSource.")
+        }
+
+        applyGeoJSONData(data: geoJSON.sourceData, sourceId: id)
+    }
+
+    // MARK: - Untyped API
+
+    internal func addSource(withId id: String, properties: [String: Any]) throws {
+        try handleExpected {
+            return styleManager.addStyleSource(forSourceId: id, properties: properties)
+        }
+    }
+
+    internal func removeSource(withId id: String) throws {
+        try handleExpected {
+            return styleManager.removeStyleSource(forSourceId: id)
+        }
+
+        workItems.removeValue(forKey: id)?.cancel()
+    }
+
+    internal func sourceExists(withId id: String) -> Bool {
+        return styleManager.styleSourceExists(forSourceId: id)
+    }
+
+    internal func sourceProperty(for sourceId: String, property: String) -> StylePropertyValue {
+        return styleManager.getStyleSourceProperty(forSourceId: sourceId, property: property)
+    }
+
+    internal func sourceProperties(for sourceId: String) throws -> [String: Any] {
+        return try handleExpected {
+            return styleManager.getStyleSourceProperties(forSourceId: sourceId)
+        }
+    }
+
+    internal func setSourceProperty(for sourceId: String, property: String, value: Any) throws {
+        try handleExpected {
+            return styleManager.setStyleSourcePropertyForSourceId(sourceId, property: property, value: value)
+        }
+    }
+
+    internal func setSourceProperties(for sourceId: String, properties: [String: Any]) throws {
+        try handleExpected {
+            return styleManager.setStyleSourcePropertiesForSourceId(sourceId, properties: properties)
+        }
+    }
+
+    // MARK: - Async GeoJSON source data parsing
+
+    private func addGeoJSONSource(_ source: GeoJSONSource, id: String) throws {
+        guard let data = source.data else {
+            try addSourceInternal(source, id: id)
+            return
+        }
+
+        if case GeoJSONSourceData.empty = data {
+            try addSourceInternal(source, id: id)
+            return
+        }
+
+        var emptySource = source
+        emptySource.data = .empty
+
+        try addSourceInternal(emptySource, id: id)
+
+        applyGeoJSONData(data: data, sourceId: id)
+    }
+
+    private func applyGeoJSONData(data: GeoJSONSourceData, sourceId id: String) {
+        workItems.removeValue(forKey: id)?.cancel()
+
+        let item = DispatchWorkItem { [weak self] in
+            if self == nil { return }
+
+            let json = try! data.toString()
+
+            self?.mainQueue.async { [weak self] in
```

Comment:
what if we did schedule task in the `mainQueue` but before closure gets executed `removeSource` for given geojson get called? then `self` inside closure is still there (as the map object is alive) but the source is already removed at that point

Reply:
Exactly! In this case there will be an attempt to set "data" for an unexisting source - we'll catch the error and log it.